C error when cloning array slice

**V version:** V 0.1.27 bd4f9d8
**OS:** Windows/Playground

**What did you do?**
```v
fn main() {
	x := ['a', 'b', 'c']
	x[1..].clone()
} 
```

**What did you expect to see?**
No error
 
**What did you see instead?**
```

C:\Users\emily\AppData\Local\Temp\v\vproto.tmp.c: In function 'compiler__new_type':
C:\Users\emily\AppData\Local\Temp\v\vproto.tmp.c:13042:35: error: lvalue required as unary '&' operand
    full_name_pieces = array_clone(&array_slice(context, 1, context.len));
                                   ^
C:\Users\emily\AppData\Local\Temp\v\vproto.tmp.c:13047:32: error: lvalue required as unary '&' operand
   context_no_pkg = array_clone(&array_slice(context, 1, context.len));
                                ^
```

--- a/vlib/v/gen/fn.v
+++ b/vlib/v/gen/fn.v
@@ -449,6 +449,7 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 			idx := (node.left as ast.IndexExpr).index
 			if idx is ast.RangeExpr {
 				// expr is arr[range].clone()
+				// use array_clone_static instead of array_clone
 				name = '${receiver_type_name}_${node.name}_static'.replace('.', '__')
 				is_range_slice = true
 			}

--- a/vlib/v/gen/fn.v
+++ b/vlib/v/gen/fn.v
@@ -438,7 +438,24 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 			g.write('*($return_type_str*)')
 		}
 	}
-	name := '${receiver_type_name}_$node.name'.replace('.', '__')
+
+	mut name := '${receiver_type_name}_$node.name'.replace('.', '__')
+	mut is_range_slice := false
+	if node.receiver_type.is_ptr() && !node.left_type.is_ptr() {
+		match node.left {
+			ast.IndexExpr {
+				idx := it.index
+				match idx {
+					ast.RangeExpr {
+						name = '${receiver_type_name}_${node.name}_static'.replace('.', '__')
+						is_range_slice = true
+					}
+					else {}
+				}
+			}
+			else {}
+		}
+	}
 	// if node.receiver_type != 0 {
 	// g.write('/*${g.typ(node.receiver_type)}*/')
 	// g.write('/*expr_type=${g.typ(node.left_type)} rec type=${g.typ(node.receiver_type)}*/')
@@ -452,7 +469,9 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 		// The receiver is a reference, but the caller provided a value
 		// Add `&` automatically.
 		// TODO same logic in call_args()
-		g.write('&')
+		if !is_range_slice {
+			g.write('&')
+		}
 	} else if !node.receiver_type.is_ptr() && node.left_type.is_ptr() && node.name != 'str' {
 		g.write('/*rec*/*')
 	}

--- a/vlib/v/gen/fn.v
+++ b/vlib/v/gen/fn.v
@@ -440,20 +440,18 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 	}
 
 	mut name := '${receiver_type_name}_$node.name'.replace('.', '__')
+	// Check if expression is: arr[a..b].clone(), arr[a..].clone()
+	// if so, then instead of calling array_clone(&array_slice(...))
+	// call array_clone_static(array_slice(...))
 	mut is_range_slice := false
 	if node.receiver_type.is_ptr() && !node.left_type.is_ptr() {
-		match node.left {
-			ast.IndexExpr {
-				idx := it.index
-				match idx {
-					ast.RangeExpr {
-						name = '${receiver_type_name}_${node.name}_static'.replace('.', '__')
-						is_range_slice = true
-					}
-					else {}
-				}
+		if node.left is ast.IndexExpr {
+			idx := (node.left as ast.IndexExpr).index
+			if idx is ast.RangeExpr {
+				// expr is arr[range].clone()
+				name = '${receiver_type_name}_${node.name}_static'.replace('.', '__')
+				is_range_slice = true
 			}
-			else {}
 		}
 	}
 	// if node.receiver_type != 0 {

--- a/vlib/v/tests/array_slice_clone_test.v
+++ b/vlib/v/tests/array_slice_clone_test.v
@@ -1,0 +1,11 @@
+fn test_array_slice_clone() {
+	arr := [1, 2, 3, 4, 5]
+	cl := arr[1..].clone()
+	assert cl == [2, 3, 4, 5]
+}
+
+fn test_array_slice_clone2() {
+	arr := [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+	cl := arr[1..].clone()[2..].clone()
+	assert cl == [4, 5, 6, 7, 8, 9, 10]
+}

